Creates sampler based MultiApp

aeslaughter opened the pull request:
This is based on changes in #9333.

This simply creates a MultiApp to spawn the stochastic runs based on a sampler. The idea will be to add transfers accordingly, the concept of which will be.

The sampler returns a vector of DenseMatrices, where each row of each matrix is a sub-application run. It is a vector b/c depending on the sample strategy there are multiple groups of sub-apps that need to run (i.e., SOBOL sampler). 

(closes #9328 #9409)

aeslaughter committed:
Re-factor distributions and samplers and add Weibull vis boost

(closes #9329)

aeslaughter committed:
Updates samplers to utilize matrix to make sampler more flexible

aeslaughter committed:
Creates sampler for Sobol variance-base sensitivity analysis

(refs #9328)

aeslaughter committed:
Remove temporary file that was added (#000)

aeslaughter committed:
Address reviewer comments

(refs #9328)

aeslaughter committed:
Updates documentation for Stocastic Tools module

(refs #9328)

aeslaughter committed:
Update index loops to use beginIndex and fix seed index

(refs #9328)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/98133/) on da0ccd8 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6bbdb336e11785cfc2bf21c4a0bdf989d67f5fc3`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9409/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9409/style.patch | git apply -v`

aeslaughter committed:
Generalize Sampler class and improve MooseRandom

Samplers simply return a vector of matrices, this allows the sampler to create a very general set of output such as needed by the Sobol sampler
(refs #9402)

aeslaughter committed:
Create Sampler based MultiApp for running stochastic simulations

(closes #9402)

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98253/) on cfe5ec3 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9409/site)

aeslaughter committed:
Update documentation for stochastic tools

(refs #9402)

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98358/) on e537684 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9409/site)

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98412/) on ec1d854 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9409/site)

permcody commented on framework/src/samplers/Sampler.C:
```diff
@@ -12,103 +12,73 @@
-  return _probability_weight;
-}
+  for (std::size_t i = 0; i < number; ++i)
+    _generator.seed(i, _seed + i);
```
This is bad idea. This almost guarantees that multiple simulations where seeds overlap will be invalid. seeds = {0, 1, **2, 3**} won't produce independent results from seeds {**2, 3**, 4, 5}. `RandomData` solved this problem with a master generator that produces seeds for slave generators.

permcody commented on modules/stochastic_tools/include/distributions/BoostDistribution.h:
```diff
@@ -0,0 +1,88 @@
+ * A class used to as a base for distributions defined by Boost.
+ *
+ * The default type is set to double to allow for derived classes to compile without Boost and
+ * trigger the mooseError in the constructor.
```
Nice design

permcody commented on modules/stochastic_tools/src/distributions/WeibullDistribution.C:
```diff
@@ -0,0 +1,28 @@
+}
+
+WeibullDistribution::WeibullDistribution(const InputParameters & parameters)
+  : BoostDistribution(parameters)
```
You are missing the template parameter here in the call to the base class. This is a common mistake that used to catch us all the time on at least one of the compilers we tested against (maybe the minimum) We've had to fix this over and over. It looks like the rules are changing with C++11 but you might as well be careful here:
https://stackoverflow.com/questions/8887864/template-base-constructor-call-in-member-initialization-list-error

permcody commented on modules/stochastic_tools/src/tests/TestDistributionPostprocessor.C:
```diff
@@ -21,37 +14,30 @@ validParams<TestDistributionPostprocessor>()
 {
-  return _distribution.cdf(_sample_value) * _distribution.pdf(_sample_value) *
-         _distribution.inverseCdf(_cdf_value) * _distribution.getRandomNumber();
+  if (_distribution_method == "pdf")
```
How often does getValue get called? It might be worth creating a class-level enum and using the `getEnum()` method on MooseEnum so that these lookups change from string comparisons to enum comparisons. Also, a switch statement might be more appropriate here (really helps solidify that this is a fixed list of options).

aeslaughter replied:
This is a test object only.

permcody commented on modules/stochastic_tools/src/vectorpostprocessors/SamplerData.C:
```diff
@@ -0,0 +1,56 @@
+void
+SamplerData::initialize()
+{
+  for (VectorPostprocessorValue * ptr : _sample_vectors)
```
maybe `auto`?

aeslaughter replied:
I am getting back on my soap box.

Why? VectorPostprocessorValue is already a typedef. Adding auto just masks that fact that that we are iterating over a vector, so then I have to go look at the header to see what type is underneath.

I just don't understand why we are going to auto so much. Sure I know we could name the iterator, like I did here, to indicate that the container is made of pointers, but that changes the code readability from using compile time errors on types to arbitrary variable names that will be easily wrong of overlooked.

We are using beginIndex because we can't be trusted to use the correct type, but we are going to rely on iterator names to make the code readable. This is all crazy to me.

I will get off the soap box now; I will change it to auto.

permcody replied:
Quite simply, people are really bad at guessing types. Nearly every single PR that goes up has at least one incorrectly assigned type. I agree that having types in the source can help with readability because they still describe intention even if they are slightly incorrect.

On the flip side, the more you use `auto` the more comfortable you get with just allowing the interfaces to do their jobs. It also makes refactoring a breeze when everyone uses `auto` as there are few places where you end up changing types. It'll take time, but I think people will get better at naming variables to describe usage (i.e. I really only care if something is numeric, or character based, or a vector. I don't really care if it's an int versus an unsigned long and this is where people make mistakes anyway).

permcody reviewed (dismissed):
This is getting really close. The seed thing really needs to be fixed. Other than that, most of my other comments are pretty minor.

aeslaughter commented on framework/include/samplers/Sampler.h:
```diff
@@ -15,71 +15,102 @@
+   * new numbers.
    */
-  virtual std::vector<Real> getSampledValues(const std::vector<std::string> & variableNames);
+  void execute();
```
No, this should not be virtual. The method users are supposed to use is "sample". The execute method is simply for saving state, I could change the name, but this method is called according to the "execute_on" parameter.

aeslaughter replied:
The comment is specific, because that is what the method does and it isn't supposed to do anything else.

aeslaughter committed:
Create a 'master' seed generator for sampler

(refs #9402)

aeslaughter committed:
Address reviewer comments

(refs #9402)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/98890/) on eb95a48 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format ff16aa1189fc6faed5c702fc7411ac99ee79ce62`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9409/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9409/style.patch | git apply -v`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98925/) on 39000e5 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9409/site)

permcody reviewed (approved):
Believe it or not, I think this is ready.
